### ISSUE TITLE
模型加载

### ISSUE THREAD

**Issue by Wzhsgsg** [LianjiaTech/BELLE#349]
Some weights of LlamaForCausalLM were not initialized from the model checkpoint at /home/fist/belle-7b and are newly initialized: ['model.layers.1.self_attn.rotary_emb.inv_freq', 'model.layers.9.self_attn.rotary_emb.inv_freq', 'model.layers.22.self_attn.rotary_emb.inv_freq', 'model.layers.24.self_attn.rotary_emb.inv_freq', 'model.layers.16.self_attn.rotary_emb.inv_freq', 'model.layers.25.self_attn.rotary_emb.inv_freq', 'model.layers.27.self_attn.rotary_emb.inv_freq', 'model.layers.4.self_attn.rotary_emb.inv_freq', 'model.layers.5.self_attn.rotary_emb.inv_freq', 'model.layers.20.self_attn.rotary_emb.inv_freq', 'model.layers.29.self_attn.rotary_emb.inv_freq', 'model.layers.26.self_attn.rotary_emb.inv_freq', 'model.layers.31.self_attn.rotary_emb.inv_freq', 'model.layers.3.self_attn.rotary_emb.inv_freq', 'model.layers.23.self_attn.rotary_emb.inv_freq', 'model.layers.17.self_attn.rotary_emb.inv_freq', 'model.layers.18.self_attn.rotary_emb.inv_freq', 'model.layers.21.self_attn.rotary_emb.inv_freq', 'model.layers.2.self_attn.rotary_emb.inv_freq', 'model.layers.13.self_attn.rotary_emb.inv_freq', 'model.layers.7.self_attn.rotary_emb.inv_freq', 'model.layers.0.self_attn.rotary_emb.inv_freq', 'model.layers.28.self_attn.rotary_emb.inv_freq', 'model.layers.6.self_attn.rotary_emb.inv_freq', 'model.layers.11.self_attn.rotary_emb.inv_freq', 'model.layers.19.self_attn.rotary_emb.inv_freq', 'model.layers.14.self_attn.rotary_emb.inv_freq', 'model.layers.30.self_attn.rotary_emb.inv_freq', 'model.layers.15.self_attn.rotary_emb.inv_freq', 'model.layers.10.self_attn.rotary_emb.inv_freq', 'model.layers.12.self_attn.rotary_emb.inv_freq', 'model.layers.8.self_attn.rotary_emb.inv_freq']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Human:
这种情况是为什么？

**Comment by xianghuisun** [comment]
> LlamaForCausalLM

您的模型应该还不是huggingface格式的llama模型， 模型权重的名字与LLamaForCausalLM类不符
您需要官方的llama权重，然后转成hf格式

**Comment by xianghuisun** [comment]
> Some weights of LlamaForCausalLM were not initialized from the model checkpoint at /home/fist/belle-7b and are newly initialized: ['model.layers.1.self_attn.rotary_emb.inv_freq', 'model.layers.9.self_attn.rotary_emb.inv_freq', 'model.layers.22.self_attn.rotary_emb.inv_freq', 'model.layers.24.self_attn.rotary_emb.inv_freq', 'model.layers.16.self_attn.rotary_emb.inv_freq', 'model.layers.25.self_attn.rotary_emb.inv_freq', 'model.layers.27.self_attn.rotary_emb.inv_freq', 'model.layers.4.self_attn.rotary_emb.inv_freq', 'model.layers.5.self_attn.rotary_emb.inv_freq', 'model.layers.20.self_attn.rotary_emb.inv_freq', 'model.layers.29.self_attn.rotary_emb.inv_freq', 'model.layers.26.self_attn.rotary_emb.inv_freq', 'model.layers.31.self_attn.rotary_emb.inv_freq', 'model.layers.3.self_attn.rotary_emb.inv_freq', 'model.layers.23.self_attn.rotary_emb.inv_freq', 'model.layers.17.self_attn.rotary_emb.inv_freq', 'model.layers.18.self_attn.rotary_emb.inv_freq', 'model.layers.21.self_attn.rotary_emb.inv_freq', 'model.layers.2.self_attn.rotary_emb.inv_freq', 'model.layers.13.self_attn.rotary_emb.inv_freq', 'model.layers.7.self_attn.rotary_emb.inv_freq', 'model.layers.0.self_attn.rotary_emb.inv_freq', 'model.layers.28.self_attn.rotary_emb.inv_freq', 'model.layers.6.self_attn.rotary_emb.inv_freq', 'model.layers.11.self_attn.rotary_emb.inv_freq', 'model.layers.19.self_attn.rotary_emb.inv_freq', 'model.layers.14.self_attn.rotary_emb.inv_freq', 'model.layers.30.self_attn.rotary_emb.inv_freq', 'model.layers.15.self_attn.rotary_emb.inv_freq', 'model.layers.10.self_attn.rotary_emb.inv_freq', 'model.layers.12.self_attn.rotary_emb.inv_freq', 'model.layers.8.self_attn.rotary_emb.inv_freq'] You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference. Human: 这种情况是为什么？

'model.layers.29.self_attn.q_proj.weight', 'model.layers.29.self_attn.k_proj.weight', 'model.layers.29.self_attn.v_proj.weight', 'model.layers.29.self_attn.o_proj.weight', 'model.layers.29.self_attn.rotary_emb.inv_freq', 'model.layers.29.mlp.gate_proj.weight', 'model.layers.29.mlp.down_proj.weight', 'model.layers.29.mlp.up_proj.weight', 'model.layers.29.input_layernorm.weight', 'model.layers.29.post_attention_layernorm.weight', 'model.layers.30.self_attn.q_proj.weight', 'model.layers.30.self_attn.k_proj.weight', 'model.layers.30.self_attn.v_proj.weight', 'model.layers.30.self_attn.o_proj.weight', 'model.layers.30.self_attn.rotary_emb.inv_freq', 'model.layers.30.mlp.gate_proj.weight', 'model.layers.30.mlp.down_proj.weight', 'model.layers.30.mlp.up_proj.weight', 'model.layers.30.input_layernorm.weight', 'model.layers.30.post_attention_layernorm.weight', 'model.layers.31.self_attn.q_proj.weight', 'model.layers.31.self_attn.k_proj.weight', 'model.layers.31.self_attn.v_proj.weight', 'model.layers.31.self_attn.o_proj.weight', 'model.layers.31.self_attn.rotary_emb.inv_freq', 'model.layers.31.mlp.gate_proj.weight', 'model.layers.31.mlp.down_proj.weight'

这才是hf格式的LLaMA模型权重的name

**Comment by Wzhsgsg** [comment]
@xianghuisun 感谢！旋转编码的话应该是固定值，再次使用随机初始化应该是没有影响的吧！我是想在微调的模型上继续微调，所以加载的model.bin。然后微调保存的模型没有带上rotary_emb.inv_freq这个东西，所以推断和重新加载的时候，报不相符的警告。

class LlamaRotaryEmbedding(torch.nn.Module):
    def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
        super().__init__()
        inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2).float().to(device) / dim))
        self.register_buffer("inv_freq", inv_freq)

        # Build here to make `torch.jit.trace` work.
        self.max_seq_len_cached = max_position_embeddings
        t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
        freqs = torch.einsum("i,j->ij", t, self.inv_freq)
        # Different from paper, but it uses a different permutation in order to obtain the same calculation
        emb = torch.cat((freqs, freqs), dim=-1)
        self.register_buffer("cos_cached", emb.cos()[None, None, :, :], persistent=False)
        self.register_buffer("sin_cached", emb.sin()[None, None, :, :], persistent=False)
只要维度不变的话，inv_freq应该是不变的。

**Comment by Lam1360** [comment]
> @xianghuisun 感谢！旋转编码的话应该是固定值，再次使用随机初始化应该是没有影响的吧！我是想在微调的模型上继续微调，所以加载的model.bin。然后微调保存的模型没有带上rotary_emb.inv_freq这个东西，所以推断和重新加载的时候，报不相符的警告。
> 
> class LlamaRotaryEmbedding(torch.nn.Module): def **init**(self, dim, max_position_embeddings=2048, base=10000, device=None): super().**init**() inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2).float().to(device) / dim)) self.register_buffer("inv_freq", inv_freq)
> 
> ```
>     # Build here to make `torch.jit.trace` work.
>     self.max_seq_len_cached = max_position_embeddings
>     t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
>     freqs = torch.einsum("i,j->ij", t, self.inv_freq)
>     # Different from paper, but it uses a different permutation in order to obtain the same calculation
>     emb = torch.cat((freqs, freqs), dim=-1)
>     self.register_buffer("cos_cached", emb.cos()[None, None, :, :], persistent=False)
>     self.register_buffer("sin_cached", emb.sin()[None, None, :, :], persistent=False)
> ```
> 
> 只要维度不变的话，inv_freq应该是不变的。

所以这个要怎么解决呀？

**Comment by wangzhonghai** [comment]
@Lam1360 对的，不用管，但是有些程序微调的时候，旋转编码也在微调，这样的不行。但是transformer中的程序是个固定值，所以不用管

**Comment by Lam1360** [comment]
@wangzhonghai 好的好的多谢

**Comment by bestpredicts** [comment]
> > LlamaForCausalLM
> 
> 您的模型应该还不是huggingface格式的llama模型， 模型权重的名字与LLamaForCausalLM类不符 您需要官方的llama权重，然后转成hf格式

check了一下.self_attn.rotary_emb.inv_freq 不在模型的可调参数model.named_parameters() 里面。但是在model.state_dict()里面。 如果用model.load_state_dict保存的模型，确实会提示找不到这个key的值，合理的

**Comment by daliqiqi** [comment]
https://github.com/huggingface/transformers/pull/24998  transformerd 版本的问题
